### DISCUSSION TITLE
Class activation maps API explained

### DISCUSSION THREAD

**Discussion by kapsner** [Project-MONAI/MONAI#1604] [Q&A] [answered]
Hi,

I do have some questions on how to correctly apply the monai class activation maps (CAM) API and would be very happy to get some help on this topic.

* a) What are the differences between `monai.visualize.class_activation_maps.CAM` and `monai.visualize.class_activation_maps.GradCAM`? 
* b) How are the CAMs to be interpreted when setting the argument `class_idx=None` vs. `class_idx=y_true`?
* c) What does the argument `layer_idx=- 1` exactly do? (I was unable to find further information on this argument in the [API documentation](https://docs.monai.io/en/latest/visualize.html#monai.visualize.class_activation_maps.GradCAM.compute_map))
* d) Which layers (across different model architectures) are typically to be chosen for the argument `target_layers`?

Many thanks in advance.

**Comment by rijobro** [comment, marked as answer]
a. for a difference between CAM and Grad-CAM, see [this](https://arxiv.org/pdf/1610.02391.pdf) paper, specifically the section starting "Most relevant to our approach is the Class Activation Mapping (CAM) approach to localization...". Grad-CAM++ has also been wrapped in MONAI, see [this](https://arxiv.org/pdf/1710.11063.pdf) paper for an explanation.

b. `class_idx=None` will use the label inferred by the model -- this may be correct or incorrect depending on how good your model is. `class_idx=y_true` will use the correct label. Both can be useful "which parts of the image were important for the model to reach a given decision?" vs. "which parts of the image were important for the model to arrive at the correct decision?"

c. if you click the "source" link in the help documentation, you get a bit more information ([link](https://docs.monai.io/en/latest/_modules/monai/visualize/class_activation_maps.html#GradCAMpp.compute_map)): "index of the target layer if there are multiple target layers". `-1` will use the network's last layer, which is the most frequently used layer for CAM/Grad-CAM (hopefully this answers your fourth question, too).

**Reply by kapsner** [reply]
@rijobro Thank you very much for your answers and the deeper insights.

Indeed, regarding a) I wasn't sure, if CAM really refers to [this paper](https://arxiv.org/pdf/1512.04150.pdf). Maybe, this could be added to the API description of `monai.visualize.class_activation_maps.CAM`?

**Comment by kapsner** [comment]
Regarding question b) from above:
I have now worked a bit with GradCAM++ and often observe, that even if images were correctly classified, GradCAMs created with `class_idx=None` look completely different compared to those generated with the provided correct labels (`class_idx=y_true`).

Example of a correctly classified image (binary classification):
*class_idx=None*:
![grafik](https://user-images.githubusercontent.com/44973495/110203188-eea09080-7e6c-11eb-9399-638905542765.png)

(same image) *class_idx=1*:
![grafik](https://user-images.githubusercontent.com/44973495/110203171-d6c90c80-7e6c-11eb-8e66-b91d43c70ca0.png)

How is this to be interpreted? I would expect no differences in both GradCAMs in case of a correct classification.

Or does it mean, that the classifier considered parts of an image as relevant that let by chance / with great uncertainty (e.g. probability near but > 0.5 in a binary classification) to the correctly class and providing the true label just helps to see, to which image part the classifier would have looked at if it has known the correct class? (Which would indicate, that the classifier is not really "good").

Or is there something else to be considered?

**Reply by rijobro** [reply]
`class_idx=None` and `class_idx=y_true` should look the same if the network infers the correct class. I notice that the slider across the top is at different points for the two visualisations. Is it possible that the difference you're seeing is because you're displaying different z-slices of the image?

If that's not the case then we're into debugging! 

1. Can you share the snippet of code you use for generating the images?
2. Is it possible that a new image is selected each time you call the GradCam, such that your top and bottom images are for two different images?
3. Can you create a subplot with the original image, the gradcam for `class_idx=None` and then the gradcam for all classes (do a for loop: `for i in num_classes: gradcam(class_idx=i, ...)`)?

**Reply by kapsner** [reply]
@rijobro thx for your answer - I think, we are into debugging - I am working on one-channel 2D images and the slider position indicates just different visualizations created for the same image.

Some background on my research task:
* I am working on a binary classification
* loss = [BCEWithLogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html)
* target-shape: one-hot encoded binary label
* I am using [pytorch lightning](https://github.com/PyTorchLightning/pytorch-lightning)
* images are logged from `test_step`
* monai-version: 0.5.dev2109 (monai-weekly)
* pytorch-lightning version: 1.2.1

Regarding your suggestions:
1) for a relevant excerpt from the code please refer to the end of this post
2) I don't believe that GradCam is calculated on different images - please see code below (before, `img2plot` was a simple list and is now a dict to fully control the plotting of each image; but switching from list to dict didn't make any difference)
3) I have implemented the loop as suggested. Notably, all 3 GradCAM images (None, class_idx=0, class_idx=1) look different.
(Providing the original image is currently not possible since I am working on medical images in a research project - if required, I can maybe try to rebuild a fully reproducible example with some publicly available data).

One new example, showing the output of a correctly classified image of class y_true = 1:

*class_idx=None*:
![grafik](https://user-images.githubusercontent.com/44973495/110238664-3a256e00-7f43-11eb-9e46-fa22aac652a7.png)

*class_idx=0*:
![grafik](https://user-images.githubusercontent.com/44973495/110238693-50332e80-7f43-11eb-8f7c-fa9511dcaaac.png)

*class_idx=1*:
![grafik](https://user-images.githubusercontent.com/44973495/110238713-6fca5700-7f43-11eb-8027-1606d45dec5d.png)


```python
import monai

import pytorch_lightning as pl
from pytorch_lightning.metrics.utils import to_categorical
import torch

from matplotlib import pyplot as plt
from skimage import color

import numpy as np


class Net(pl.LightningModule):
    # https://pytorch.org/tutorials/beginner/blitz/neural_networks_tutorial.html
    def __init__(self, conf):
        super().__init__()

        self.save_hyperparameters(conf)

        # load model architecture:
        self.model = monai.networks.nets.densenet.densenet121(
            pretrained=self.hparams.pretrained,
            spatial_dims=2,
            in_channels=1,
            out_channels=self.hparams.num_classes,
            bn_size=self.hparams.bn_size,
            dropout_prob=self.hparams.dropout_prob
        )

        # instantiate monai transforms to log misclassifies test images
        self.img_resize = monai.transforms.Compose([
            monai.transforms.Resize(spatial_size=[256, 256])
        ])
        self.img_rescale = monai.transforms.Compose([
            monai.transforms.ScaleIntensity(minv=0, maxv=1)
        ])

        # set test_img num
        self.test_img_num = 0

        # set test iteration
        self.test_iteration = 0

        # instantiate class activation map function
        self.cam_models = [
            "densenet121"
        ]

        if self.hparams.model_name in [
            "densenet121"
        ]:
            target_layer = "class_layers.relu"

        # instantiate class activation maps only if current model in cam_models
        if self.hparams.model_name in self.cam_models:

            self.gradcampp = monai.visualize.GradCAMpp(
                nn_module=self.model,
                target_layers=target_layer
            )

    def forward(self, x):
        return self.model(x)

    def test_step(self, batch, batch_idx, prefix):
        x = batch["image"]
        y = batch["target"]

        logits = self(x)

        # prepare targets
        y_preds, y_trues = self._y_to_cat(
            logits=logits,
            targets=y
        )

        self._log_images(
            data=x,
            y_preds=y_preds,
            y_trues=y_trues,
            ids=batch["id"]
        )

    def _log_images(self, data, y_preds, y_trues, ids):

        y_preds = y_preds.cpu().detach().numpy()
        y_trues = y_trues.cpu().detach().numpy()

        for raw_image, _y_pred, _y_true, _id in zip(data, y_preds, y_trues, ids):

            # create output-text
            tag = "y_true: {0} -- y_pred: {1} (test iteration {2}) / {3} - ID {4}".format(
                _y_true,
                _y_pred,
                self.test_iteration,
                self.test_img_num,
                _id
            )
            cm_images = {}

            if self.hparams.model_name in self.cam_models:

                with torch.set_grad_enabled(True):

                    # GradCAM++
                    # without class_idx
                    cm_images["None"] = self.gradcampp(
                        x=raw_image[None],
                        class_idx=None
                    )

                    # with class_idx
                    for _cls in range(self.hparams.num_classes):
                        cm_images[_cls] = self.gradcampp(
                            x=raw_image[None],
                            class_idx=_cls
                        )

            # dict of images to plot later
            img2plot = {}

            # get colormap
            cm = plt.get_cmap("jet_r")

            # iterate over images that require color map
            for _img_key, _img_val in cm_images.items():
                # apply colormap
                _img = cm(np.squeeze(_img_val))
                # create 3-channel image for displaying in tensorboard
                img2plot[_img_key] = np.transpose(
                    color.rgba2rgb(_img),
                    (2, 0, 1)
                )

            # process original image
            # add axis, required for transformations (c, h, w)
            img = np.squeeze(raw_image.cpu().detach().numpy())[None]
            # rescale original image
            img = self.img_rescale(img)  # intensity=0-1, c, h, w
            # create 3-channel grayscale image for displaying in tensorboard
            img2plot["original_image"] = np.transpose(
                color.gray2rgb(np.squeeze(img)),
                (2, 0, 1)
            )

            img_keys = ["None"] + \
                [i for i in range(self.hparams.num_classes)] + \
                ["original_image"]
            for _index, _img in enumerate(img_keys):
                _img = self.img_resize(img2plot[_img])  # c, h, w
                monai.visualize.img2tensorboard.plot_2d_or_3d_image(
                    data=_img[None],
                    step=_index + 1,
                    writer=self.logger[0].experiment,
                    max_channels=3,
                    max_frames=1,
                    tag=tag
                )

            # increment image number
            self.test_img_num += 1

    @staticmethod
    def _y_to_cat(logits, targets):
        y_preds = to_categorical(
            tensor=torch.sigmoid(logits),
            argmax_dim=1
        )
        y_trues = to_categorical(
            tensor=targets,
            argmax_dim=1
        )
        return y_preds, y_trues
```

**Reply by kapsner** [reply]
@rijobro after writing the post, it just came to me that the issue could be related to the use of monai-weekly from pip.

So I just tested the code again with monai 0.4.0 and can confirm that the error does not occur here.
There seems to be an issue with / a bug in monai-weekly....

Gradcams for the same image as above with monai 0.4.0: correctly classified, y_true = 1.
As expected, gradcams 1 (class_idx=None) and 3 (class_idx=y_true=1) look exactly the same.

*class_idx=None*:
![grafik](https://user-images.githubusercontent.com/44973495/110239363-da30c680-7f46-11eb-910f-a8b42375e0d2.png)

*class_idx=0*:
![grafik](https://user-images.githubusercontent.com/44973495/110239419-119f7300-7f47-11eb-84cd-973a11abd123.png)

*class_idx=1*:
![grafik](https://user-images.githubusercontent.com/44973495/110239439-211ebc00-7f47-11eb-99f6-bb422e99447d.png)

**Reply by rijobro** [reply]
Ok, thanks for bringing this to our attention, I'll create an issue for it. If you wanted to help with debugging, it would be really useful to create a minimum (not)working example. If it were using a pre-trained densenet, then we wouldn't even need to do any training to see the issue.

**Reply by rijobro** [reply]
Issue created: https://github.com/Project-MONAI/MONAI/issues/1710

**Reply by kapsner** [reply]
Thanks a lot - I will see if I can provide a minimum working example soon!

**Reply by rijobro** [reply]
@kapsner - do you think you could try this PR and let me know if it solves your problem? https://github.com/Project-MONAI/MONAI/pull/1734

**Reply by kapsner** [reply]
@rijobro I appologize for not yet providing a working example - didn't find enough time yet.

However, I can confirm, that this PR solves the problem for me! Thank you very much!!!
